torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -347,16 +356,44 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
   /// The base case, when the list of modules is empty.
   void push_back() {}
 
-  // Box the AnyModules to give Sequential reference semantics, like the rest of
-  // the API. Note that this is not required otherwise, this could just be a
+  // Box the AnyModules to give AnySequential reference semantics, like the rest
+  // of the API. Note that this is not required otherwise, this could just be a
   // `vector<AnyModule>`.
   std::vector<AnyModule> modules_;
 };
 
+/// A `ModuleHolder` subclass for `AnySequentialImpl`.
+/// See the documentation for `AnySequentialImpl` class to learn what methods it
+/// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+/// module storage semantics.
+TORCH_MODULE(AnySequential);
+
+/// `Sequential` is derived from `AnySequential` and has a non-template
+/// returning type allowing `Sequential`s to be nested in each other making them
+/// compatible with the python API. The template input type helps warn users
+/// about the change in the API.
+class SequentialImpl : public AnySequentialImpl {
+ public:
+  using AnySequentialImpl::AnySequentialImpl;
+
+  template <typename InputType = Tensor>
```

Comment:
We are actually constraining to output type to be Tensor here, not the input type.